### ISSUE TITLE
Early adopters can experiment with native AOT form factor

### ISSUE THREAD

**Issue by samsp-msft** [dotnet/runtimelab#248]
_This is a draft/proposed User Story and not committed for .NET 6.0 yet._

We have seen a surge of interest in statically linked binaries with minimal dependencies in recent years. This is a reversal of a multi-decade trend to deliver dependencies as shared libraries. The benefits to deployment and the ability to optimize a single app outweigh the benefits of sharing for certain types of workloads. The gain is most significant for workloads with a high number of deployed instances: cloud infrastructure, hyper-scale services, popular apps or games. The number of apps of this type is relatively small, but they are highly visible, impactful, and often cost a lot of money to operate (savings can be meaningful).

Emerging programming environments (e.g. Go, Rust) tend to be designed for this form factor. Established programming environments are catching up on this trend to be relevant for this highly visible segment, for example Java Graal, Dart, Kotlin/Native.

Various implementations of .NET (Xamarin, Unity, .NET Native) have demonstrated that .NET is well suited to native AOT in terms of fundamentals, however, there are tools experience and ecosystem challenges that currently prevent .NET from being a great general purpose solution. We intend to start addressing these issues with this project. It is very likely that the first set of improvements will have rough edges that may prevent adoption for some developers. We will use feedback to determine which experience issues to prioritize. We are also not comitting to any particular end point for this project, at this time.

This high level section sets our intention to invest in [native AOT scenarios](https://github.com/dotnet/designs/blob/main/accepted/2020/form-factors.md#native-aot-form-factors) and more definition from the teams will be needed to define our phased approach over the next few .NET releases.

# Work Items / User stories

- [x] Source Generators should be created to replace existing reflection heavy code with build-time source generation so that it can be used in a native AOT environment. 
  - [x] System.Text.Json serialization – code generators can work around the need for runtime reflection, and generate serializers for specific classes. https://github.com/dotnet/runtime/issues/1568
  - [ ] EventSource – event source uses reflection.emit to simplify creation of classes (manifests) for custom event sources – this can be changed to a build-time activity using source generators. https://github.com/dotnet/runtime/issues/43390
- [x] Invest in wider platform support and debugging support for Native AOT https://github.com/dotnet/runtimelab/issues/336
- [ ] Developers can precompile their Regex code for faster startup https://github.com/dotnet/runtime/issues/44676
 + TBD

**Comment by masonwheeler** [comment]
Be careful to define this properly.  A "JIT-less native AOT" is not a good idea.  Far too much of the .NET ecosystem depends on the JIT being there for one reason or another.  A far better goal is to AOT everything that can be determined statically, and then leave the JIT in there for when it's needed.  It doesn't have to be an either/or thing.  Just another example of Anders Hejlsberg's principle: static where possible, dynamic when necessary.

**Comment by PJB3005** [comment]
@masonwheeler certain platforms like e.g. iOS don't allow JITs, I believe. Supporting a JITless scenario is the only option there.

I believe the "AOT and JIT at once" scenario is already filled by ReadyToRun, no?

**Comment by masonwheeler** [comment]
> @masonwheeler certain platforms like e.g. iOS don't allow JITs, I believe.

Then we shouldn't support them.  Particularly iOS; why legitimize a system that's spent over a decade being actively hostile to you?  Like a person in an abusive relationship, attempting to appease them never makes anything better; the only correct choice of action is to walk away.

**Comment by RalfKornmannEnvision** [comment]
> > @masonwheeler certain platforms like e.g. iOS don't allow JITs, I believe.
> 
> Then we shouldn't support them. 

This would cut the option to support game consoles, too. As far as I remember there were quite some interest for native AOT from game developers in the survey.

**Comment by PJB3005** [comment]
> This would cut the option to support game consoles,

Indeed, I believe that none of the game console platforms allow JITs either (though of course detailed info on this kinda stuff will be behind NDAs).

There's a reason Mono AOT is necessary, and also Unity's IL2CPP.

> Then we shouldn't support them. Particularly iOS; why legitimize a system that's spent over a decade being actively hostile to you? Like a person in an abusive relationship, attempting to appease them never makes anything better; the only correct choice of action is to walk away.

Not getting into any of the "abusive relationship" stuff... not supporting iOS is a huge market missed and also means Android is less viable (no ability to share code). That's not practical.

> Not really. It doesn't go anywhere near far enough in AOTing things that can be AOT'd; as soon as a problem starts to look even a little bit difficult it punts it off to the JIT. I say, aggressively JIT everything that can be JITted, and leave the AOT only for things that legitimately require it.

Did you mix up "AOT" and "JIT" in your last sentence? At first you say it AOTs too little but then you say it should JIT as much as possible.

**Comment by masonwheeler** [comment]
@RalfKornmannEnvision Which game console forbids JITs on their platform?  I've only ever heard of Apple doing that with iOS.  (It's quite possible that I haven't heard about it on consoles because of NDAs or whatever, but the fact remains that I haven't heard about it happening anywhere other than iOS.  Do they have a problem with JITs in general, or just with runtime generation of executable code in-memory, and if the latter, could the IL interpreter be employed?)

**Comment by masonwheeler** [comment]
@PJB3005 
> Did you mix up "AOT" and "JIT" in your last sentence? At first you say it AOTs too little but then you say it should JIT as much as possible.

Oops, you're right.  Fixed now.

**Comment by EgorBo** [comment]
Is there a list of things which R2R/JIT aren't able to precompile? (does JIT support/use shared generics?)

**Comment by RUSshy** [comment]
If i want AOT compilation, i don't want a JIT compiler in my program, by definition

JIT = non deterministic performance, it is same story as GC, if i make game, i don't want random stutters

I don't understand why people want to force the JIT into people's program

**Comment by RalfKornmannEnvision** [comment]
> @RalfKornmannEnvision Which game console forbids JITs on their platform? I've only ever heard of Apple doing that with iOS. (It's quite possible that I haven't heard about it on consoles because of NDAs or whatever, but the fact remains that I haven't heard about it happening anywhere other than iOS. Do they have a problem with JITs in general, or just with runtime generation of executable code in-memory, and if the latter, could the IL interpreter be employed?)

Every modern game console doesn't allow to generate any kind of executable code at runtime. It's part of the security system. 

You can always run an interpreter but on a system like the Nintendo Switch I would honestly prefer to have an option to not depend on it even if it means that I lost some features.Developers who already use one of the C# solutions for these platforms are already aware of such limitations. But using RyuJit instead of Mo0no AOT would mean an significant performances improvement.

**Comment by masonwheeler** [comment]
> I don't understand why people want to force the JIT into people's program

To cover the edge cases.  If you have code that's using dynamic code generation (expression trees, serializers, etc,) calling reflection features such as `MakeGenericType`, or using recursive generics, these are things that simply cannot be evaluated at compile-time because they depend on information that isn't available until runtime.

**Comment by RUSshy** [comment]
> > I don't understand why people want to force the JIT into people's program
> 
> To cover the edge cases. If you have code that's using dynamic code generation (expression trees, serializers, etc,) calling reflection features such as `MakeGenericType`, or using recursive generics, these are things that simply cannot be evaluated at compile-time because they depend on information that isn't available until runtime.

for edge cases, edge solutions, it should be opt-in, not opt-out, default should cover what people expect when they AOT compile, not more, not less

**Comment by PJB3005** [comment]
These kinds of edge cases are better solved in ways such as better compile time code generation, which we are getting with C# source generators for example.

**Comment by masonwheeler** [comment]
> for edge cases, edge solutions, it should be opt-in, not opt-out, default should cover what people expect when they AOT compile

What people expect when they use .NET code is *.NET code that works.*  That should always be the highest priority.

**Comment by PJB3005** [comment]
> What people expect when they use .NET code is .NET code that works. That should always be the highest priority.

I fail to see how this is mutually exclusive?

**Comment by masonwheeler** [comment]
> These kinds of edge cases are better solved in ways such as better compile time code generation, which we are getting with C# source generators for example.

Oh, I definitely agree.  And if source generators were actually any good, that would be amazing.  But they aren't.  The current implementation of generators really sucks, and the Roslyn team is firmly committed to making sure they never stop being sucky, because good metaprogramming is incompatible with something they hold as a much higher value: ensuring that everything will execute very quickly and never make Intellisense have to wait.  (Not exaggerating.  This is the result of many exasperating conversations with team members who are completely unwilling to budge on this point.  Visual Studio UX trumps language quality any day of the week.)

**Comment by RUSshy** [comment]
> > These kinds of edge cases are better solved in ways such as better compile time code generation, which we are getting with C# source generators for example.
> 
> Oh, I definitely agree. And if source generators were actually any good, that would be amazing. But they aren't. The current implementation of generators really sucks, and the Roslyn team is firmly committed to making sure they never stop being sucky, because good metaprogramming is incompatible with something they hold as a much higher value: ensuring that everything will execute very quickly and never make Intellisense have to wait. (Not exaggerating. This is the result of many exasperating conversations with team members who are completely unwilling to budge on this point. Visual Studio UX trumps language quality any day of the week.)

I agree 100% with that, i even made a suggestion here that i closed because the team doesn't seems to like "hard" tasks https://github.com/dotnet/csharplang/issues/2939

Source Generators is a cheap solution and it's not even good, just like XAML, more bloat

> What people expect when they use .NET code is .NET code that works. That should always be the highest priority.

My project is the proof that .NET code works with CoreRT, even when reflection fully disabled

What people do with their code is their problem, they should be educated when they use LINQ and abuse generics, or emit code at runtime etc etc etc

The runtime needs to be cleaned too

Other languages doesn't have this issue, and we see more and more "high level" system languages, many of them proved that these solutions made at runtime can be done at compile time, and be much cleaner and more efficient

**Comment by RalfKornmannEnvision** [comment]
> What people do with their code is their problem, they should be educated when they use LINQ and abuse generics, or emit code at runtime etc etc etc

I don't think that this would be a big problem. Developers are already using Xamarin on iOS, Unity or Mono AOT today All of these have limitations compared to the full .Net experiences  

But even if there are already working solutions for AOT I still like to see am offical supported solution that is based on the current .net development. Because based on own experiments I know that the RuyJit based CoreRT solution generates much better results than Mono AOT.

**Comment by masonwheeler** [comment]
> Other languages doesn't have this issue, and we see more and more "high level" system languages, many of them proved that these solutions made at runtime can be done at compile time, and be much cleaner and more efficient

Rust does it by having actual metaprogramming, of the type the Roslyn folks refuse to consider.  Not sure which other languages you're referring to.  The only other "high level system language" that's gained any traction in recent years is Go, and Go is basically awful all around in this area.

**Comment by HurricanKai** [comment]
> Oh, I definitely agree.  And if source generators were actually any good, that would be amazing.  But they aren't.  The current implementation of generators really sucks, and the Roslyn team is firmly committed to making sure they never stop being sucky, because good metaprogramming is incompatible with something they hold as a much higher value: ensuring that everything will execute very quickly and never make Intellisense have to wait.  (Not exaggerating.  This is the result of many exasperating conversations with team members who are completely unwilling to budge on this point.  Visual Studio UX trumps language quality any day of the week.)

The current version of source generators isn't great, that's true. But it'll be improved. And imo it's very important we keep the fast nature of Roslyn. Do we really want another C++ with a build process more complicated then your code, and longer compilation time then Runtime? (exaggeration, of course, but you get the point) 

Roslyn is very very powerful and it's not at its limits with source generators at all, theres lots of improvements to be made imo, but I don't doubt they'll be made for one second. 

Also, I believe the only thing the Roslyn team refuses to consider is code removal/rewriting. Not sure what everyone thinks this will do for them, but to my knowledge another core principle of the Roslyn code base is immutability, and this makes sense from a language design perspective too, I've never heard a single compelling case for code rewriting. There are already security concerns with analyzers, and even more of them with source generators...

**Comment by masonwheeler** [comment]
> Also, I believe the only thing the Roslyn team refuses to consider is code removal/rewriting.

And given that rewriting is the single most useful feature of metaprogramming, my complaint stands.

**Comment by pjmlp** [comment]
Thanks for reconsidering the status of AOT support.

**Comment by zakbaig** [comment]
Would be absolutely amazing if Blazor switches to using NativeAOT (assuming it uses CoreRT as its base). CoreRT produced much smaller binaries than Mono’s AOT effort while still producing best in class performance. It would solve the question of “can I use Blazor for a massive public facing website?” as AOT Wasm is extremely fast so there’d be no performance penalty of using Blazor over JavaScript SPA frameworks, and the start-up time would, in theory, only be slightly slower than these aforementioned frameworks.

I’m convinced that if Microsoft want Blazor to be the go-to answer for web development then it needs to use the gold mine that is CoreRT. There’s only a few too many uncomfortable trade-offs a person has to make when using today’s Blazor but they can be eliminated entirely with a good NativeAOT/CoreRT story.

**Comment by TimPosey2** [comment]
I'd also add in that (LLILC)[https://github.com/dotnet/llilc] could pair nicely with a native AOT implementation to target multiple backends and optimizations via LLVM. Auto-vectorization comes to mind.

**Comment by RalfKornmannEnvision** [comment]
> I'd also add in that [LLILC ](https://github.com/dotnet/llilc) could pair nicely with a native AOT implementation to target multiple backends and optimizations via LLVM. Auto-vectorization comes to mind.

The issue with LLILC is that you will lost all the special net optimization's that are done by RyuJit. Because of this there is the idea to to to use a combination of RyuJit and LLVM. First let RyuJit convert and optimize the MSIL code to the CPU independent internal RyuJit intermediate format. Than instead of generating code for the target CPU the RyuJit format can be converted to LLVM. This will allow future optimization's that are beyond the limitations of RyuJit. This way you can get the best of both worlds. But this is even further away than NativeAOT.

**Comment by EgorBo** [comment]
I have a hand-made demo where at some phase of RyuJIT (lowering) I emit LLVM IR (using LLVM C++ API) from GenTrees/BBs 🙂 
works for very simple methods, initially I wanted to create a tier2/3 LLVM JIT demo 🙂 
Just like @RalfKornmannEnvision said, it allows me to emit LLVM IR after JIT made all the .NET-specific optimizations: de-boxing, range-check elimination, etc.

**Comment by HurricanKai** [comment]
> And given that rewriting is the single most useful feature of metaprogramming, my complaint stands.
I've not ever heard a good argument for code rewriting. I'd appreciate if you gave one. @masonwheeler

**Comment by masonwheeler** [comment]
> I've not ever heard a good argument for code rewriting. I'd appreciate if you gave one. @masonwheeler

Have you personally done any metaprogramming work, or are you just relying on arguments for/against that you hear from others?  Because when you're actually doing it yourself, you run into rewriting scenarios *all the time.*

My go-to example is parsing.  With metaprogramming, you can write "source code" that declaratively defines a grammar, and use a rewriter to turn it into imperative code that parses that grammar.  I have actually done this, multiple times.  One time I even went back later and tweaked the rewriter so that it would not only generate a parser but also a serializer, so that data files in the esoteric binary format I was parsing could be read, edited, and then saved again.

Being able to separate the low-level parsing principles from the specific grammar rules of the entities I was parsing made the task about two orders of magnitude easier than trying to write the parser and serializer logic manually would have been.

**Comment by HurricanKai** [comment]
Well I've worked with source generators a lot, and never felt rewriting was a needed feature.

What you describe can already be done, just not within a single method. The old "source code" will still be in the assembly. The user could just specify what methods to use (and declare them as partial / just make you create them from scratch).
The only bit you need rewriting for is the user magically calling your generated function. Imo the understandability and security downsides far outweight the difficulties of asking a user to add a simple "MyGeneratedFunction()"

**Comment by masonwheeler** [comment]
Another highly beneficial thing that can be done with rewriters (and is quite impossible without them) is detecting eagerly evaluated LINQ-to-objects chains and rewriting them as foreach loops by substituting each LINQ method call with the equivalent behavior.  (Again, a thing I have actually done.  It produces some serious performance gains in hot-path code.)

**Comment by tannergooding** [comment]
> Another highly beneficial thing that can be done with rewriters (and is quite impossible without them) is detecting eagerly evaluated LINQ-to-objects chains and rewriting them as foreach loops by substituting each LINQ method call with the equivalent behavior. (Again, a thing I have actually done. It produces some serious performance gains in hot-path code.)

It's definitely not impossible, just more work and leaves an unused method.

You can define something like:
```csharp
[LinqTemplateMethod(nameof(TemplateMethod))]
public partial void PublicMethod();

private void TemplateMethod()
{
    // All the logic
}
```

You then find all methods with the attribute, analyze the code in the specified method (`TemplateMethod`), and use that to generate the improved code for `PublicMethod`.
The only "downside" is that `TemplateMethod` sticks around and is basically "dead code" for this particular scenario.

However, in other scenarios (like auto-vectorization), it can be a software fallback that is used if the hardware doesn't support the "improved" path.

**Comment by tannergooding** [comment]
Its definitely not as "ideal" as being able to replace arbitrary code inline, but you can still opt-in and get all the benefits as needed more explicitly. It also avoids having multiple potential C# dialects or having to do 2+ pass compilations to ensure the existing code is all valid C# as well as the newly generated code and provides an all around better IDE experience, which is what 90% of users will have anyways.

**Comment by samsp-msft** [comment]
The conversations about code-rewriting are interesting, but are quite a bit off-track from this original issue 😃, perhaps they should continue over in the [roslyn repo](https://github.com/dotnet/roslyn/issues?q=is%3Aissue+label%3A%22New+Feature+-+Source+Generators%22+)?

I am wondering if there could be an extension to partial methods to allow for replacement provided the signature of the replacement matches.

**Comment by masonwheeler** [comment]
> > Another highly beneficial thing that can be done with rewriters (and is quite impossible without them) is detecting eagerly evaluated LINQ-to-objects chains and rewriting them as foreach loops by substituting each LINQ method call with the equivalent behavior. (Again, a thing I have actually done. It produces some serious performance gains in hot-path code.)
> 
> It's definitely not impossible, just more work and leaves an unused method.

"It's not impossible to do this; you just have to work around the impossibility by doing something completely different that takes more work and will produce similar results except for this unavoidable drawback over here" does not actually demonstrate that the thing in question is not impossible.

**Comment by trampster** [comment]
There is no unavoidable drawback, CoreRT/NativeAOT is very good a stripping unused code.

**Comment by TheSpydog** [comment]
If I'm reading the edit history correctly, over the past few days this issue has transformed from a concrete statement of intent and a plan of action toward making NativeAOT a first-class .NET scenario, into a nebulous, low-priority proposal that may or may not be accepted, with no work items associated.

This does not instill much hope or confidence in this project, which is deeply disappointing. It felt like there was finally some momentum toward an official embrace of AOT, given the recent survey, the migration of CoreRT to the runtimelab repo, and the original wording of this issue. But apparently that's not the case after all?

**Comment by jkotas** [comment]
We are doing planning for .NET 6 on github in public. You can watch live and comment on “how the sausage is made”.
- This issue is linked from https://dotnetepics.azurewebsites.net/ that has the live global view of the .NET wide epics. This is work in progress and the set will change a lot before we are done with the planning. However, you can see how this fits into overall plan from that view.
- This issue is marked as “Priority 0” currently. “Priority 0” is the highest priority. It means that we believe that funding this is very important.
- The wording above reflects the fact that native AOT for .NET is a very complex problem that we expect to work on over multiple releases.

**Comment by TheSpydog** [comment]
Thank you for the clarification! That's very good to hear.

> * This issue is marked as “Priority 0” currently. “Priority 0” is the highest priority.

This is... an interesting labeling scheme. :) From an outside perspective, it's ambiguous as to whether this is the lowest or highest priority level. IMHO it might be worth considering changing this label to something more indicative of its "highest priority" status.

> * The wording above reflects the fact that native AOT for .NET is a very complex problem that we expect to work on over multiple releases.

Of course, the wording is very clear about that. However, the removal of the long-term, high-level action items from the original wording of the OP and the addition of a "draft/proposal" disclaimer (combined with the lowest/highest priority ambiguity) seemed to indicate a sudden retreat from committing to AOT, hence my previous confusion and disappointment. I am very glad to hear definitively that this is not the case.

**Comment by danmosemsft** [comment]
@TheSpydog for whatever reason, the Priority 0 through Priority 3 scheme has been commonly used within Microsoft for a long time so it's "what we know". It would be reasonable to ask whether we should do something different when its public, though. Something to consider next time we go through this kind of planning process.

**Comment by HurricanKai** [comment]
GitHub labels can have a description, could add one to the Priority labels.

**Comment by andrew-boyarshin** [comment]
Chromium and Mozilla use the same priority label naming scheme, so it's not Microsoft-specific and pretty well-understood. But label description wouldn't hurt either.

**Comment by samsp-msft** [comment]
> GitHub labels can have a description, could add one to the Priority labels.

done.

As the person responsible for the original text and most of the edits - they have mostly been made to match the internal planning processes. I cleared out the work items as the original list was incomplete, and will be re-added as we iterate further in the planning process.
What you are seeing in github are snapshots "as the sausage is being made", and you are not necessarily seeing the internal processes that force some of the updates. The issue churn in github will also be a bit higher as we are also transitioning from using Azure DevOps to using github issues, and are evolving the process as we go.

**Comment by webczat** [comment]
as for adding a jit for dynamic cases, is there any option of having an opt-in feature, to add the jit in case I need the dynamic features for any reason, for example... idk... loading plugins? That could be disabled by default and it would be understood it doesn't always work like context of ios/game consoles. Also adding a jit will make the binary grow so the developer will get to decide whether he wants to just compile to IL and run with jit, or just AOT, or AOT and have the JIT somehow included for dynamic loading of stuff or something like that... Not sure how much it makes sense or not. Or you'd expect for this case people wouldn't need AOT at all?

**Comment by jkotas** [comment]
The existing shipping .NET Core/.NET 5 is optimized for apps that need dynamic features like loading plugins. It comes with R2R that is AOT tech designed for this scenario.

This issue is specially about Native AOT defined [here](https://github.com/dotnet/designs/blob/main/accepted/2020/form-factors.md#native-aot-form-factors).

**Comment by webczat** [comment]
well why does native AOT preclude additionally linking in an optional jit? I mean for these platforms that do allow it. I mean what is technically preventing it to be mixed in case the developer wants it? othervise if they still want a main singlefile app it would be way larger...

**Comment by masonwheeler** [comment]
> The existing shipping .NET Core/.NET 5 is optimized for apps that need dynamic features like loading plugins. It comes with R2R that is AOT tech designed for this scenario.
> 
> This issue is specially about Native AOT defined [here](https://github.com/dotnet/designs/blob/main/accepted/2020/form-factors.md#native-aot-form-factors).

The point is that the Native AOT definition is incomplete.  There are too many things in the CLR ecosystem that rely on runtime code generation for one purpose or another.  You can get 95% of the way without ever needing it, but then you run face-first into the lack of a JIT trying to do the last 5%.  Having a JIT available to fall back on as a last resort would make Native AOT an order of magnitude easier to work with and more likely to find adoption more quickly.

Remember Anders Hejlsberg's principle: *static where possible, dynamic when needed.*  Typing is not the only valid application of this concept.

**Comment by jkotas** [comment]
> There are too many things in the CLR ecosystem that rely on runtime code generation for one purpose or another

Yes, Native AOT is not for those things. It is clearly stated in the definition "Be explicit about that this form factor is not a good fit for every application out there.".

You may not like the definition, but it does not make it incomplete.

**Comment by jkotas** [comment]
> well why does native AOT preclude additionally linking in an optional jit?

Because of then you end up with something that is super close to CoreCLR. There is no point in building two different solutions that have very similar characteristics.

**Comment by masonwheeler** [comment]
> > There are too many things in the CLR ecosystem that rely on runtime code generation for one purpose or another
> 
> Yes, Native AOT is not for those things. It is clearly stated in the definition "Be explicit about that this form factor is not a good fit for every application out there.".
> 
> You may not like the definition, but it does not make it incomplete.

The fact that it's incomplete is what makes me not like it.  It's not "not a good fit for every application out there"; *it's a poor fit for virtually everything* because that last 5% is ubiquitous.  Just to name the most obvious example, who's going to want to build a .NET application where they can't even get basic serialization to work without jumping through all sorts of painful hoops?

**Comment by trampster** [comment]
@masonwheeler Painful hoops? I do json Serialization with NativeAOT using  JsonSrcGen which is a source generator based Json Lib that doesn't require Reflection.Emit or JIT.  It doesn't require any painful hoops.

All the major use cases are already covered by Source Generators. And this is only going to get better with time as more are developed.

**Comment by jkotas** [comment]
We know about important customers that really want the Native AOT form factor as it is defined, that will abandon .NET if it does not materialize, that are fine with the limitations and have no problem of doing extra work to e.g. hook up serialization generation at build time. Part of this epic is source generator for System.Text.Json that make this even easier (you can only get it via 3rd party serializers like Utf8Json today).

This is also the reasons why we are running this as an experimental project in runtimelab to prove that this can make these customers successful. (I am sorry, but I cannot share the customer names.)

If you believe that the Native AOT form factor is not going to work for you, you can safely ignore it. As state in https://github.com/dotnet/designs/blob/main/accepted/2020/form-factors.md#continuous-improvements , our primary focus is to improve the existing successful form factors, so you do not need to worry about being left out.

**Comment by masonwheeler** [comment]
@trampster Source Generators *are* a painful hoop!  C#'s implementation is horribly inadequate because the language team's stubbornness in prioritizing top-tier Visual Studio responsiveness *uber alles* impedes them from creating a metaprogramming solution that's actually any good.

**Comment by flibitijibibo** [comment]
From our perspective on the FNA project, the JIT would actually be quite problematic if it were even an option, as it adds a pretty significant amount of ambiguity for portability (which is what AOT's use case is for game development in the first place). Typically you want either all-JIT or all-AOT, anything in the middle is just giving you the same amount of work that you get from transitioning between the two, but now you're stuck having to untangle things if someone does end up wiring the JIT into an AOT environment where JITs are explicitly not allowed (read: every single console and mobile environment).

Serialization and dynamic loading are not a zero-step process to port, but under absolutely no circumstances have we found this to be insurmountable, nor have we found that the solution that was ultimately developed was not worth the development costs, even without factoring that _the whole reason that AOT for .NET exists is because JITs are explicitly banned in most environments_. jkotas can't name names, but I can, so a couple real-world examples from me:

- For FNA games like Dust: An Elysian Tail and Solaroids, the most common trip-up is the ContentManager - XNA Content depends pretty heavily on .NET serialization and reflection, so the expectation is that you'll load the type info dynamically from the managed assemblies and it'll all "just work". This is not true in console environments, but BRUTE has a config file for exactly this issue - you explicitly list the content types that would otherwise get missed by the type spider and it works perfectly fine after that. Ignoring that JITs are banned to begin with, a single config file is a lot easier to maintain than a JIT plugin in this case. At most it's about a day of work if you got a little too carried away with templates.
- For reflection-heavy FNA games like River City Ransom: Underground and Streets of Rage 4, all the game logic risks being _extremely_ dependent on what a JIT can do, and an AOT would fall flat on its face. Rather than wedge in a JIT, however, both studios independently made their own IL emitter that writes all that information at authoring time, making it more usable in environments like CoreRT. It took some work, but it also turns out that it works well in a JIT environment too since you can get a lot of busy work done at startup instead of letting jitters happen in the middle of gameplay (a similar issue is graphics pipeline caches and shaders; you may have seen modern games do a "first-start" load to compile all shaders to avoid stuttering during gameplay).

If you want both, target both separately. The likelihood of the two targets being any different even after jumping through AOT hoops is, in our experience, always zero by the end. To put it plainly: We would not trust developers to use a JIT plugin responsibly.

**Comment by webczat** [comment]
trying to use a JIT plugin in environments where it's not supported could just fail hard at build time. What I am concerned about is that aoted binaries will be smaller than the current self contained approach, but you cannot have both smaller binaries and ability to dynamically load external IL code. Unless some form of dynamic loading is actually considered, but plugins would be AOT too? is that even doable or necessary?

**Comment by MichalStrehovsky** [comment]
> What I am concerned about is that aoted binaries will be smaller than the current self contained approach, but you cannot have both smaller binaries and ability to dynamically load external IL code.

Smaller binary implies trimming (i.e. removal of unused code). If we trim the main app and then load a plugin, we can easily end up in a situation where code that the plugin needs has been removed.

If the set of plugins is known at compile time, we might as well AOT compile the plugin too.

If the set of plugins is not known at compile time, no trimming can be performed. For Native AOT, this means we need to carry the entire framework, including stuff like DTD support in XML, FTP client, and System.Drawing.

You can't get "small & plugin support" at the same time. You actually get penalized in full AOT because native code tends to be bigger than IL - the app having a lot of unreferenced code hurts size a lot more in native AOT.

**Comment by webczat** [comment]
Oh well, I get the point. Like you cannot really know what the plugin will want to use and you cannot also dynamically load random framewowrk stuff that way... and then basically go back to jitting.
As for size it's interesting, somehow I thought it is the metadata that takes the most space, but it was always mostly a guess. Also in aot you still have the metadata don't you? as in reflection actually still works, it's just trimming that may strip stuff... I am curious how large would a possible aoted helloworld binary be in the native aot setting.

**Comment by MichalStrehovsky** [comment]
> As for size it's interesting, somehow I thought it is the metadata that takes the most space, but it was always mostly a guess. Also in aot you still have the metadata don't you? as in reflection actually still works, it's just trimming that may strip stuff... I am curious how large would a possible aoted helloworld binary be in the native aot setting.

Most of the metadata becomes optional in Native AOT. Things like names of methods, method parameters, signatures of methods, etc. are only carried around for the purposes of reflection. Runtime doesn't need it to run code.

You can grab the sample from https://github.com/dotnet/runtimelab/tree/feature/NativeAOT/samples/HelloWorld and then play with the switches documented at https://aka.ms/OptimizeCoreRT.

A fully self-contained Hello world on Windows is about 4-5 MB by default right now because the AOT compiler tries to maximize compatibility at the expense of everything else. With all switches turned on, a self-contained Hello World is under 1 MB. I expect we'll be able to end up closer to the 1 MB mark _by default_ once all the things we have in flight fall into place and we don't have to do heuristics to keep what we assume is needed (even though it's not).

**Comment by webczat** [comment]
do you expect most new apps becoming AOT when the feature is introduced? As in, when some specific app doesn't need features like dynamic loading/unloading or stuff like that, and there are no compatibility or whatever concerns, and no platform-specific requirements that prevent jitting (like mobile apps), what drives the decision between aot and jit? aot by default unless something that only jit provides is needed, or jit by default unless you have requirements that make aot better suited?

**Comment by jkotas** [comment]
I do not expect that we would ever turn on Native AOT by default. [`PublishReadyToRun`](https://docs.microsoft.com/en-us/dotnet/core/deploying/ready-to-run) (the existing shipping AOT tech) is not turned on by default either.

We believe that majority of .NET developers are happy with the current experience (fast build times, not worrying about which features are available, etc.) that turning on AOT by default would regress.

**Comment by charlesroddie** [comment]
[COM interop](https://github.com/dotnet/runtimelab/issues/306) needs to be added because it's the most impactful thing tracked in this repo, and it would fix a major regression in the ecosystem by replacing netnative. System.Text.Json and EventSource source generators and precompiling regex code are nice-to-have features rather than essentials, while COM interop opens up frameworks where native AOT can immediately get real usage, with large benefit for developers and users.

The title of this issue is not ambitious enough because early adopters can already experiment with native AOT form factor, and they have been able to experiment with it and even use it in production for years! It was better before: "Developers can compile .NET apps into native executables (native AOT form factor)".

**Comment by wjk** [comment]
My opinion on the matter: I, today, can take a .NET 5 Windows Forms project, run it though NativeAOT, and have the window appear as expected. (When you try to do anything further you get a crash due to missing COM support.) However, these days I use WPF. WPF uses C++/CLI in two of its assemblies: DirectWriteForwarder and ReachFramework. I have never used ReachFramework (hell, I don't even know what it *does*), but DirectWriteForwarder is an integral WPF component, very near the root of the dependency graph.

You can create a lot of crazy MSIL code using C++/CLI. I fully understand that it is impossible to support every possible C++/CLI assembly out there. However, not supporting *any* C++/CLI immediately rules out usage of WPF with NativeAOT. This would be a dealbreaker for me, as I (and many others) write 100% WPF applications because WPF is the only actually usable C# GUI toolkit available. (Windows Forms looks extremely dated, and WinUI 3 is nowhere near usable yet.)

**Comment by hez2010** [comment]
> WPF is the only actually usable C# GUI toolkit available. (Windows Forms looks extremely dated, and WinUI 3 is nowhere near usable yet.)

You may try out Avalonia (https://github.com/AvaloniaUI/Avalonia), which is a WPF inspired, native-aot compatible and cross-platform UI framework.

**Comment by wjk** [comment]
@hez2010 I actually have already extensively experimented with that technology, but thanks for the pointer!

**Comment by cyraid** [comment]
> > WPF is the only actually usable C# GUI toolkit available. (Windows Forms looks extremely dated, and WinUI 3 is nowhere near usable yet.)
> 
> @wjk You may try out [Avalonia](https://github.com/AvaloniaUI/Avalonia), which is a WPF inspired, native-aot compatible and cross-platform UI framework, providing almost the same experience with WPF.

Seems like it's only setup for Mac OSX Native right now? (looking at the source)

**Comment by ThomasFOG** [comment]
> From our perspective on the FNA project, the JIT would actually be quite problematic if it were even an option, as it adds a pretty significant amount of ambiguity for portability (which is what AOT's use case is for game development in the first place). Typically you want either all-JIT or all-AOT, anything in the middle is just giving you the same amount of work that you get from transitioning between the two, but now you're stuck having to untangle things if someone does end up wiring the JIT into an AOT environment where JITs are explicitly not allowed (read: every single console and mobile environment).
> 
> Serialization and dynamic loading are not a zero-step process to port, but under absolutely no circumstances have we found this to be insurmountable, nor have we found that the solution that was ultimately developed was not worth the development costs, even without factoring that _the whole reason that AOT for .NET exists is because JITs are explicitly banned in most environments_. jkotas can't name names, but I can, so a couple real-world examples from me:
> 
> * For FNA games like Dust: An Elysian Tail and Solaroids, the most common trip-up is the ContentManager - XNA Content depends pretty heavily on .NET serialization and reflection, so the expectation is that you'll load the type info dynamically from the managed assemblies and it'll all "just work". This is not true in console environments, but BRUTE has a config file for exactly this issue - you explicitly list the content types that would otherwise get missed by the type spider and it works perfectly fine after that. Ignoring that JITs are banned to begin with, a single config file is a lot easier to maintain than a JIT plugin in this case. At most it's about a day of work if you got a little too carried away with templates.
> * For reflection-heavy FNA games like River City Ransom: Underground and Streets of Rage 4, all the game logic risks being _extremely_ dependent on what a JIT can do, and an AOT would fall flat on its face. Rather than wedge in a JIT, however, both studios independently made their own IL emitter that writes all that information at authoring time, making it more usable in environments like CoreRT. It took some work, but it also turns out that it works well in a JIT environment too since you can get a lot of busy work done at startup instead of letting jitters happen in the middle of gameplay (a similar issue is graphics pipeline caches and shaders; you may have seen modern games do a "first-start" load to compile all shaders to avoid stuttering during gameplay).
> 
> If you want both, target both separately. The likelihood of the two targets being any different even after jumping through AOT hoops is, in our experience, always zero by the end. To put it plainly: We would not trust developers to use a JIT plugin responsibly.

I would like to back Ethan's comment as a MonoGame contributor and in regard to my position at Flying Oak Games (we make, and port games to consoles).

NativeAOT, as it is now, has been an extremely valuable project and several relevant projects have been shipped with it to wide audiences. This includes Streets of Rage 4, Carrion, ScourgeBringer, River City Ransom: Underground, Dust... which all shipped to Windows, macOS, and Linux, and to more than 100k players each.
The tech is clearly appreciated for its efficient and bloat-free approach of AOT compilation. We can clearly say that it has been battle-tested for games. Even third-party providers like Epic or GOG have shown us good support and were keen to update their own C# API to be specifically NativeAOT-friendly.

I would also very much back the counterproductivity of AOT+JIT in that context. C# game developers are very used to work without a JIT (they want to avoid it at all cost because it is visible during gameplay, and a forced AOT is an excellent benchmark to proof a code to work without JIT) and without reflection because most console runtimes don't support it in favor of aggressive AOT. As mentioned by Ethan, game developers are more into pre-generating serializers (e.g. with protobuf) than relying on runtime IL emition (this is a trend that I can clearly confirm).

In regard to reflection, BRUTE (the AOT compiler used for FNA and MonoGame on consoles) is using a very straightforward xml description file to let the compiler know which types should be included if they were to be trimmed by spidering. It works great as it is.

TL;DR; NativeAOT is a very relevant, appreciated and battle-tested project as it is for game development, and I would very much support its integration as an official .NET form-factor.

**Comment by ryancheung** [comment]
I'm also making a mmorpg game that have both client and server written with dotnet 6 and built with NativeAOT. NativeAOT provides more efficient code, and most importantly for me, it protects my source code almost as c++ would do.

**Comment by agocke** [comment]
We've hit our targets here for 6.0 -- new architecture support (ARM64 on Linux and Windows) and many bugs fixed. Huge amounts of netcoreapp libraries have been annotated for trimming (https://github.com/dotnet/runtime/issues/43543) which is the most important indicator of NativeAOT compatibility. Many libraries have even been made safe for trimming, like the source generators for System.Text.Json (https://github.com/dotnet/runtime/issues/1568). We've also started adding more CoreCLR feature support to NativeAOT, like default interface implementations and covariant returns.

This is just the start of NativeAOT work, but .NET 6 has tons of improvements. We'll soon start planning for .NET 7 and we'll see what else we can do for NativeAOT and how the improvements can fit into the broader .NET 7 release.

**Comment by zezba9000** [comment]
Have any experiments been done that just compile (C# => LLVM => ASM) OR (C# => ASM) directly?
In that you skip IL completely. One benefit langs designed around AOT runtimes is more of the runtime & implementation details can be updated faster which means performance issues caused from legacy IL baggage don't become blockers for basic features. Or one lang isn't limited because of the limitations of other langs.

Another thought: If a new C++ feature comes out, older compilers can't parse the headers & thats fine. Same concept should apply to .NET binaries.  Seems like its far to often newer C# features take forever to be considered because of older runtime compatibility nonsense. This really needs to be mitigated more and newer runtimes designed to handle newer lang features in performant manors should be the focus. Thats a hard Pro over any minor Cons IMO. This would allow features like fixed multi-dimensional arrays & fixed non-primitive arrays to be added very quickly I would imagine.

AOT has valuable lessons that seem to be getting missed in the .NET / JIT world IMO.

**Comment by sgf** [comment]
.net 6 has been release,but seems both  (AOT or Native) not included.

**Comment by mxashlynn** [comment]
If anyone has the time, a brief summary of the state of AOT in .Net 6 including how early adopters can begin experimenting with it would be greatly appreciated.  🙂

**Comment by jkotas** [comment]
> Have any experiments been done that just compile (C# => LLVM => ASM) OR (C# => ASM) directly?

While these experiments can be interesting research projects, they would lead towards a complete reset and fork of .NET ecosystem. We do not believe that it would be a viable strategy. See ingredients of successful native AOT support in https://github.com/dotnet/designs/blob/main/accepted/2020/form-factors.md#native-aot-form-factors

> C# features take forever to be considered because of older runtime compatibility nonsense. 

It used to to be the case with .NET Framework. We are well pass that. We have been adding new core runtime features in every .NET Core release that are not present in older runtimes.

**Comment by jkotas** [comment]
> .net 6 has been release,but seems both (AOT or Native) not included.

.NET had different forms of AOT since forever. https://docs.microsoft.com/en-us/dotnet/core/deploying/ready-to-run is form of AOT for server and desktop included in .NET 6. There is also AOT for wasm and mobile form factors.

> how early adopters can begin experimenting with it would be greatly appreciated. 

NativeAOT packages are published on experimental feed. It is enough for early adopters that we have been in contact with to experiment with it. See https://github.com/dotnet/runtimelab/blob/feature/NativeAOT/docs/using-nativeaot/compiling.md for how to use it.

**Comment by zezba9000** [comment]
> It used to to be the case with .NET Framework. We are well pass that. We have been adding new core runtime features in every .NET Core release that are not present in older runtimes.

-- This is only partially true. Which is why I specifically brought up "fixed multi-dimensional arrays & fixed non-primitive arrays" which are being blocked by MSIL & legacy runtime compatibility issues with said IL (which is what needs to be mitigated). And thats just the one I know of because I've wished this part of the lang was feature complete for years. I'm sure their are others in this category.

**Comment by jkotas** [comment]
> "fixed multi-dimensional arrays & fixed non-primitive arrays" which are being blocked by MSIL & legacy runtime compatibility issues with said IL

fixed arrays are a type system concern. IL is just one of the many places where the .NET type system is materialized. Other places include language compilers, AOT compilers, runtime, reflection APIs, debuggers, and all sort of other tools. Adding new core concept requires touching most of these. The problem is updating all these places, not in updating the IL alone.

**Comment by zezba9000** [comment]
> fixed arrays are a type system concern... The problem is updating all these places, not in updating the IL alone.
* I understand but this is true for any feature and the type / object system in the C# lang is already there for this as its recognized just not compilable to existing IL as it would seem by the kind of errors you get.

* The "fixed-array" example illustrates how the IL / runtime methodology the .NET team is using artificially limits runtime / IL updates on sadly non-feature-complete areas for the sake of legacy runtime compatibility (which is a artificial limitation many AOT langs like C++ don't have in this specific area)... and my thinking is if the .NET team would worry less about C# compiler compatibility so it can target older .NETFW 4.x runtimes etc this wouldn't be an issue. A .NET DLL is more like a C++ header with a .lib/.dll where the JIT compiles & links .obj files. IMO .NET JIT should be considered in a similar light to C++ compilers in terms of what is allowed to be considered breaking changes in its meta-data (in C++ terms its header file).

* Not to be redundant but IMO again you can't compile against modern C++ with old C++ compilers and the same should be true for .NET & its JIT (compiler)... if that makes sense.

**Comment by jkotas** [comment]
> my thinking is if the .NET team would worry less about C# compiler compatibility so it can target older .NETFW 4.x runtimes etc this wouldn't be an issue

I respectfully disagree. Maintaining C# compiler compatibility so it can target older .NETFW 4.x runtimes adds less than 0.1% overhead to everything we do today.

> you can't compile against modern C++ with old C++ compilers and the same should be true for .NET & its JIT (compiler)

That's true for .NET today. You cannot run modern .NET 6 C# with old .NET Framework. The opposite works though: you can compile and run old C# code with latest .NET 6. Just like it is with C++.

**Comment by zezba9000** [comment]
> Maintaining C# compiler compatibility so it can target older .NETFW 4.x runtimes adds less than 0.1% overhead to everything we do today.

But not always in the most optimal/performant way.
.NET assemblies behave like native ABIs in their virtual machine but IMO feels like this is going down the path of saying ASM never needs to change because the architecture it runs on will never change. I don't think thats inline with reality. I think long run thats a failed concept & while .NET has done great work it can start to show after awhile.

Anyway don't mean to drag this on forever...

**Comment by PJB3005** [comment]
> Have any experiments been done that just compile (C# => LLVM => ASM) OR (C# => ASM) directly?

Unity's Burst is basically an alternative optimized compiler as far as I know, and I think it still just takes in IL. (IIRC) In a blog post they mentioned that re-using the (relatively) high level IL emitted by the C# compiler was much easier than making a completely alternative C# compiler (or language altogether). Of course it's still a specialized case, but it's something to consider.

> This would allow features like fixed multi-dimensional arrays & fixed non-primitive arrays to be added very quickly I would imagine.

I don't know about non-primitive arrays ,but my understanding is that the runtime barely even knows what `fixed` is, at least from me looking at the IL output by the C# compiler. It sounds like multi-dim arrays would be totally *possible* since it's mostly C# side. (compiler just emits an internal type that is the right length for the fixed buffer and translates accesses to it, right? Can just have the compiler support multi-dim stuff)

**Comment by masonwheeler** [comment]
Multi-dimensional arrays already exist.  According to [the CLI spec:](https://www.ecma-international.org/wp-content/uploads/ECMA-335_6th_edition_june_2012.pdf)

> I.8.9.1 Array types 
> An array type shall be defined by specifying the element type of the array, the rank (number of
dimensions) of the array, and the upper and lower bounds of each dimension of the array

This is supported in the C# language through [multidimensional array syntax.](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/arrays/multidimensional-arrays)

**Comment by PJB3005** [comment]
I believe @zezba9000 is referring to `fixed` arrays instead like:
```cs
struct Foo
{
	public fixed byte Bar[32];
}
```

**Comment by jkotas** [comment]
> It sounds like multi-dim arrays would be totally possible since it's mostly C# side.

Yes, for fixed arrays of primitive types only. If you would like to have fixed arrays of non-primitive types, allow fixed array to show up in signatures, or access them via reflection, there is some level of runtime support required.

See https://github.com/dotnet/runtime/issues/61135 and related issues. This topic is best discussed there.

**Comment by zezba9000** [comment]
@PJB3005 I was referring to these incomplete lang features. (None of which compile atm)

```csharp
[StructLayout(LayoutKind.Sequential)]
struct MyStruct
{
  // some fields
}

[StructLayout(LayoutKind.Sequential)]
unsafe struct Foo
{
 public fixed MyStruct Items[123];// single-dimensional fixed-array with custom-struct
 public fixed MyStruct Items[123][28];// multi-dimensional fixed-array with custom-struct
 public fixed float Items[123][28];// multi-dimensional fixed-array with primitives
}
```

**Comment by sgf** [comment]
for (custom struct) fixed-sized-buffers 
https://github.com/dotnet/csharplang/issues/1314
https://github.com/dotnet/csharplang/blob/main/proposals/fixed-sized-buffers.md

.net 6 -C#10 ,no processing yet .
